src/Analyzers/Analyzers/src/StartupFacts.cs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Analyzers
+{
+    internal static class StartupFacts
+    {
+        public static bool IsStartupClass(StartupSymbols symbols, INamedTypeSymbol type)
+        {
+            if (symbols == null)
+            {
+                throw new ArgumentNullException(nameof(symbols));
+            }
+
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            // It's not good enough to just look for a method called ConfigureServices or Configure as a hueristic.
+            // ConfigureServices might not appear in trivial cases, and Configure might be named ConfigureDevelopment
+            // or something similar.
+            //
+            // Additionally, a startup class could be called anything and wired up explicitly.
+            //
+            // Since we already are analyzing the symbol it should be cheap to do a pass over the members.
+            var members = type.GetMembers();
+            for (var i = 0; i < members.Length; i++)
+            {
+                if (members[i] is IMethodSymbol method && (IsConfigureServices(symbols, method) || IsConfigure(symbols, method)))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        // Based on StartupLoader. The philosophy is that we want to do analysis only on things
+        // that would be recognized as a ConfigureServices method to avoid false positives.
+        //
+        // The ConfigureServices method follows the naming pattern `Configure{Environment?}Services` (ignoring case).
+        // The ConfigureServices method must be public.
+        // The ConfigureServices method can be instance or static.
+        // The ConfigureServices method cannot have other parameters besides IServiceCollection.
+        //
+        // The ConfigureServices method does not actually need to accept IServiceCollection
+        // but we exclude that case because a ConfigureServices method that doesn't accept an
+        // IServiceCollection can't do anything interesting to analysis.
+        public static bool IsConfigureServices(StartupSymbols symbols, IMethodSymbol symbol)
```

Comment:
@Tratcher - any cool thoughts?

Reply:
I'm making the simplifying assumption that this analyzer won't work with the pattern where you return an `IServiceProvider`. The analyzers are for 3.0+